### ISSUE TITLE
Maximum call stack size exceeded when serializing yxmltext to DOM or JSON in afterTransaction event

### ISSUE THREAD

**Issue by NilSet** [yjs/yjs#506]
**Checklist**

* [x] Are you reporting a bug? Use github issues for bug reports and feature requests. For general questions, please use https://discuss.yjs.dev/
* [x] Try to report your issue in the correct repository. Yjs consists of many modules. When in doubt, report it to https://github.com/yjs/yjs/issues/

**Describe the bug**
When serializing a YXmlText element into its textual content, yjs internally creates a cleanup transaction to try and clean up any invalid markup which may have arisen when merging updates from two different agents. Since 13.5.43, if there is currently an active transaction, this cleanup transaction is added to the list of transactionCleanups as a followup transaction. In the case where we are serializing an xmlnode which contains thousands of these YXmlText elements, THOUSANDS OF TRANSACTIONS get added to the transactionCleanups list. In the process of processing this list thru recursion, the stack is overflowed with the message `RangeError: Maximum call stack size exceeded`

**To Reproduce**
```
  const ydoc = new Y.Doc()
  const yxml = ydoc.getXmlFragment('')
  ydoc.on('afterTransaction', tr => {
    if (tr.origin === 'test') {
      yxml.toJSON()
    }
  })
  ydoc.transact(tr => {
    for (let i = 0; i < 15000; i++) {
      yxml.push([new Y.XmlText('a')])
    }
  }, 'test')
  ```

**Expected behavior**
I see three distinct problems with how the system is behaving currently.
1. transaction side effects from reading the textual contents of a YXmlText is very surprising and has now proven harmful
2. letting the empty transactions from these pessimistic cleanup checks pile up in transactionCleanups due to the active transaction not matching is a problem. 
3. using recursion to process a list of unknown length when working thru the transactionCleanups leads to stack overflows. Iteration would probably not have the same problem.

**Environment Information**
 - any, threshold for the bug may change based on how much stack is allocated for your env
 - Yjs >= 13.5.43

**Additional context**
We asked for this transaction behavior and it turns out its even more broken for our use case oops.

**Comment by dmonad** [comment]
We can make the transaction within toJSON optional. The transaction is only needed when reading from a snapshot. That should fix the problem. But I don't know yet how this should be handled correctly in case you want to read a snapshot within a `afterTransaction` handler.

**Comment by NilSet** [comment]
@dmonad could you explain in more detail what exactly the [transaction inside YText.toDelta](https://github.com/yjs/yjs/blob/7e40fc442d4b2f3e527de20fef40ddddc0ee3caf/src/types/YText.js#L1021-L1082) is needed for? Does it ever actually make changes to the document?

**Comment by dmonad** [comment]
Hi @NilSet,

Thanks so much for this issue. Sorry I just now fixed it.

It's actually quite simple. Calling `toJSON` on Y.Xml will generate one cleanup transaction for every single `Y.XmlText` object.

The cleanup transaction in `Y.Text` is only necessary for optimizing the "struct store" after calling `toDelta(snapshot)` with an optional snapshot parameter. In order to revisit an older version (snapshot) we need to perform some changes on the "struct store". We revert them once the delta is computed.

Another reason (this is more relevant to v14) to create a "cleanup transaction" is to optimize the `Y.Text` data structure and remove duplicate formatting attributes. E.g. `"{bold-start}hello {bold-start} world{bold-end}"` - the second `bold-start` is not necessary.

This issue is easily fixed by wrapping `toJSON` inside a transaction. Then all cleanup transactions will reuse the existing transaction object.

Thank you for the PR #507! In general, we shouldn't create too many small transaction objects. So if this occurs again, we should wrap the respective method in a transaction again. The cleanup process is quite expensive, especially if you have to do it 15k times. In a sense, I'm happy that a stack overflow is thrown because we wouldn't have caught it otherwise. There should always only be a few transaction cleanups. I might reject your PR just because of this "feature".

**Comment by dmonad** [comment]
FYI: Just published a fix in 13.5.49 and added your code as a test-case.

**Comment by dmonad** [comment]
@SamyPesse 
This is probably a different issue. Could you please give steps to reproduce it?

In your case, the issue seems to be thrown in the `update` event.

**Comment by SamyPesse** [comment]
Here is a repository to repro: https://github.com/SamyPesse/yjs-repro

It contains a `update.bin` which is a snapshot of an update that can create this situation. I've extracted part of the rest of our logic where we try to convert the `Y.Doc` into a different data structure.

You can run with `node test.js` (you need to install `yjs` next to it).

Thank you!

**Comment by dmonad** [comment]
Hi @SamyPesse,

You can avoid the problem by wrapping [`YChildrenToNodes`](https://github.com/SamyPesse/yjs-repro/blob/main/test.js#L26) in a transaction. E.g.:

```javascript
function yChildrenToNodes(element) {
    return Y.transact(element.doc, () => element.toArray().map((node) => {
        if (node instanceof Y.XmlElement) {
            return yElementToNode(node);
        } else if (node instanceof Y.XmlText) {
            return yTextToText(node);
        } else {
            throw new Error(`Unknown node type: ${node}`);
        }
    }), 'cleanup');
}
```

This is how I fixed the issue in `toJSON`..

However, this is excellent feedback! Manually iterating through the document (using `toArray`, ..) is a common use case. Calling `toDelta` on children should be possible without wrapping everything inside transactions, which is quite counterintuitive. So I decided that `toDelta` shouldn't create cleanup transactions at all, when not necessary. 

The issue should be fixed for you as well in `Yjs@v13.5.50`.

**Comment by SamyPesse** [comment]
Amazing! thank you @dmonad

**Comment by fsctl** [comment]
Hi @dmonad , I just opened PR #523 (and associated issue #522) for the same underlying problem.  My PR fixes the issue the same way (change recursive function into an iterative one), but it does so by **changing only four lines of code**, so it's very easy to see that it's logically equivalent to the current code.

[You commented above](https://github.com/yjs/yjs/issues/506#issuecomment-1462011540) that you're inclined to reject PRs that fix the recursion issue because you view it as a "feature" in some ways.  

I understand where you're coming from, but I'd really like to encourage you to reconsider.  These call stack overflows make Y.js unusable for large documents, and it was a pain for us to debug this and get to the root of the problem.  And then just today I found a Stack Overflow post where someone else is having exactly the same issue ([here](https://stackoverflow.com/questions/75931668/data-persistence-due-to-rangeerror-maximum-call-stack-size-exceeded-y-js)).  Bad experiences like this cause people to give up on Y.js -- my company almost did.  

I think it would benefit the project to change that function to be iterative.  If you like, it could instead emit a warning on the console that the transaction count was too large instead of crashing the whole app.  Would that be a good middle ground?
